### DISCUSSION TITLE
Sealed interface

### DISCUSSION THREAD

**Discussion by wrexbe** [dotnet/csharplang#7451] [Language Ideas]
public sealed interface IMyInterface
{
    
}

It is a public interface, that can be used in other assemblies, but not implemented outside of the assembly it is defined in.

**Comment by bernd5** [comment]
I would not see "sealed interface" as can't be implentend in foreign assemblies. For me it looks that it can not be extended by other interface declarations.
What is your reason to prohibit that?

**Reply by wrexbe** [reply]
Public APIs that want to expose an interface, but do not want to worry about users implementing it. One area of use cases is implementing the Memento design pattern.

**Comment by HaloFour** [comment]
See: #486

**Reply by wrexbe** [reply]
Yes, basically this, it's easier to achieve something with classes, but not with interfaces

**Comment by wrexbe** [comment]
Maybe the interface can specify the types allowed to implement it with attributes in addition to the sealed

[ImplementedBy(typeof(something)]

**Comment by jnm2** [comment]
Existing approaches would be:

1. Ship an analyzer along with your interface to block it, like the compiler does: https://github.com/dotnet/roslyn/blob/main/src/Compilers/Core/Portable/Symbols/ISymbol.cs#L23C6-L23C32

2. Put `[Obsolete("Implementing this interface is not supported.", DiagnosticId = "ABC123")]` on the interface, and then globally silence that diagnostic ID within the the assembly that may implement that interface.

You can work around these if you're determined, but the messaging can make it clear that you'll only hurt yourself with that kind of determination.
